helix-term/src/commands/typed.rs

```diff
@@ -2167,6 +2169,64 @@ fn reset_diff_change(
     Ok(())
 }
 
+fn rename_buffer(
+    cx: &mut compositor::Context,
+    args: &[Cow<str>],
+    event: PromptEvent,
+) -> anyhow::Result<()> {
+    if event != PromptEvent::Validate {
+        return Ok(());
+    }
+    ensure!(args.len() == 1, ":rename takes one argument");
+    let new_name = args.first().unwrap();
+
+    let (_, doc) = current!(cx.editor);
+    let path = doc
+        .relative_path()
+        .ok_or_else(|| anyhow!("Buffer not saved, use :write"))?
+        .canonicalize()
+        .map_err(|_| anyhow!("Could not get absolute path of the document"))?;
+    let mut path_new = path.clone();
+    path_new.set_file_name(OsStr::new(new_name.as_ref()));
+    if path_new.exists() {
+        bail!("This file already exists");
+    }
+
+    if let Err(e) = std::fs::rename(&path, &path_new) {
+        bail!("Could not rename file, error: {}", e);
+    }
+    let (_, doc) = current!(cx.editor);
+    doc.set_path(Some(path_new.as_path()))
+        .map_err(|_| anyhow!("File renamed, but could not set path of the document"))?;
+
+    if let Some(lsp_client) = doc.language_server() {
```

Comment:
the wii rename request must be sent (and processed) *before* the rename happens so this needs to be moved to the start of the function:

> The will rename files request is sent from the client to the server before files are actually renamed as long as the rename is triggered from within the client either by a user action or by applying a workspace edit. The request can return a WorkspaceEdit which will be applied to workspace before the files are renamed. 

At the end of the function you must send didRename:

> The did rename files notification is sent from the client to the server when files were renamed from within the client.

you also need to support mullti LSP here. Probably ok to just send wiiRename to the first server that supports it and didRename to all of them